File: api/gateway/_native/src/api/book.ts
import { Request } from 'express'
import { getGrpcError } from '~/lib/grpc-exception'
import { getGrpcMetadata } from '~/lib/grpc-metadata'
import { bookClient } from '~/plugins/grpc'
import {
  BookListResponse,
  BookResponse,
  BookshelfListResponse,
  BookshelfResponse,
  CreateBookRequest,
  DeleteBookshelfRequest,
  EmptyBook,
  GetBookByIsbnRequest,
  GetBookRequest,
  GetBookshelfRequest,
  GetReviewRequest,
  ListBookByBookIdsRequest,
  ListBookReviewRequest,
  ListBookshelfRequest,
  ListUserReviewRequest,
  ReadBookshelfRequest,
  ReadingBookshelfRequest,
  ReleaseBookshelfRequest,
  ReviewListResponse,
  ReviewResponse,
  StackBookshelfRequest,
  UpdateBookRequest,
  WantBookshelfRequest,
} from '~/proto/book_apiv1_pb'
import {
  IBookInputAuthor,
  ICreateBookInput,
  IDeleteBookshelfInput,
  IGetBookByIsbnInput,
  IGetBookInput,
  IGetBookshelfInput,
  IGetReviewInput,
  IListBookByBookIdsInput,
  IListBookReviewInput,
  IListBookshelfInput,
  IListUserReviewInput,
  IReadBookshelfInput,
  IReadingBookshelfInput,
  IReleaseBookshelfInput,
  IStackBookshelfInput,
  IUpdateBookInput,
  IWantBookshelfInput,
} from '~/types/input'
import {
  IBookHashOutput,
  IBookHashOutputAuthor,
  IBookHashOutputBook,
  IBookOutput,
  IBookOutputAuthor,
  IBookshelfListOutput,
  IBookshelfListOutputAuthor,
  IBookshelfListOutputBook,
  IBookshelfListOutputBookshelf,
  IBookshelfOutput,
  IBookshelfOutputAuthor,
  IBookshelfOutputBook,
  IBookshelfOutputReview,
  IReviewListOutput,
  IReviewListOutputOrder,
  IReviewListOutputReview,
  IReviewOutput,
} from '~/types/output'

export function listBookByBookIds(req: Request<any>, input: IListBookByBookIdsInput): Promise<IBookHashOutput> {
  const request = new ListBookByBookIdsRequest()
  const metadata = getGrpcMetadata(req)

  request.setBookIdsList(input.bookIds)

  return new Promise((resolve: (res: IBookHashOutput) => void, reject: (reason: Error) => void) => {
    bookClient.listBookByBookIds(request, metadata, (err: any, res: BookListResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      resolve(setBookHashOutput(res))
    })
  })
}

export function listBookshelf(req: Request<any>, input: IListBookshelfInput): Promise<IBookshelfListOutput> {
  const request = new ListBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setLimit(input.limit)
  request.setOffset(input.offset)

  return new Promise((resolve: (output: IBookshelfListOutput) => void, reject: (reason: Error) => void) => {
    bookClient.listBookshelf(request, metadata, (err: any, res: BookshelfListResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookshelfListOutput = setBookshelfListOutput(res)
      resolve(output)
    })
  })
}

export function listUserReview(req: Request<any>, input: IListUserReviewInput): Promise<IReviewListOutput> {
  const request = new ListUserReviewRequest()
  const metadata = getGrpcMetadata(req)

  const order = new ListUserReviewRequest.Order()
  order.setBy(input.by)
  order.setDirection(input.direction)

  request.setUserId(input.userId)
  request.setLimit(input.limit)
  request.setOffset(input.offset)
  request.setOrder(order)

  return new Promise((resolve: (output: IReviewListOutput) => void, reject: (reason: Error) => void) => {
    bookClient.listUserReview(request, metadata, (err: any, res: ReviewListResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IReviewListOutput = setReviewListOutput(res)
      resolve(output)
    })
  })
}

export function listBookReview(req: Request<any>, input: IListBookReviewInput): Promise<IReviewListOutput> {
  const request = new ListBookReviewRequest()
  const metadata = getGrpcMetadata(req)

  const order = new ListBookReviewRequest.Order()
  order.setBy(input.by)
  order.setDirection(input.direction)

  request.setBookId(input.bookId)
  request.setLimit(input.limit)
  request.setOffset(input.offset)
  request.setOrder(order)

  return new Promise((resolve: (output: IReviewListOutput) => void, reject: (reason: Error) => void) => {
    bookClient.listBookReview(request, metadata, (err: any, res: ReviewListResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IReviewListOutput = setReviewListOutput(res)
      resolve(output)
    })
  })
}

export function getBook(req: Request<any>, input: IGetBookInput): Promise<IBookOutput> {
  const request = new GetBookRequest()
  const metadata = getGrpcMetadata(req)

  request.setId(input.bookId)

  return new Promise((resolve: (output: IBookOutput) => void, reject: (reason: Error) => void) => {
    bookClient.getBook(request, metadata, (err: any, res: BookResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookOutput = setBookOutput(res)
      resolve(output)
    })
  })
}

export function getBookByIsbn(req: Request<any>, input: IGetBookByIsbnInput): Promise<IBookOutput> {
  const request = new GetBookByIsbnRequest()
  const metadata = getGrpcMetadata(req)

  request.setIsbn(input.isbn)

  return new Promise((resolve: (output: IBookOutput) => void, reject: (reason: Error) => void) => {
    bookClient.getBookByIsbn(request, metadata, (err: any, res: BookResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookOutput = setBookOutput(res)
      resolve(output)
    })
  })
}

export function getBookshelf(req: Request<any>, input: IGetBookshelfInput): Promise<IBookshelfOutput> {
  const request = new GetBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setBookId(input.bookId)

  return new Promise((resolve: (output: IBookshelfOutput) => void, reject: (reason: Error) => void) => {
    bookClient.getBookshelf(request, metadata, (err: any, res: BookshelfResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookshelfOutput = setBookshelfOutput(res)
      resolve(output)
    })
  })
}

export function getReview(req: Request<any>, input: IGetReviewInput): Promise<IReviewOutput> {
  const request = new GetReviewRequest()
  const metadata = getGrpcMetadata(req)

  request.setReviewId(input.reviewId)

  return new Promise((resolve: (output: IReviewOutput) => void, reject: (reason: Error) => void) => {
    bookClient.getReview(request, metadata, (err: any, res: ReviewResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IReviewOutput = setReviewOutput(res)
      resolve(output)
    })
  })
}

export function createBook(req: Request<any>, input: ICreateBookInput): Promise<IBookOutput> {
  const request = new CreateBookRequest()
  const metadata = getGrpcMetadata(req)

  const authors = input.authors?.map((item: IBookInputAuthor) => {
    const author = new CreateBookRequest.Author()

    author.setName(item.name)
    author.setNameKana(item.nameKana)

    return author
  })

  request.setTitle(input.title)
  request.setTitleKana(input.titleKana)
  request.setDescription(input.description)
  request.setIsbn(input.isbn)
  request.setPublisher(input.publisher)
  request.setPublishedOn(input.publishedOn)
  request.setThumbnailUrl(input.thumbnailUrl)
  request.setRakutenUrl(input.rakutenUrl)
  request.setRakutenSize(input.rakutenSize)
  request.setRakutenGenreId(input.rakutenGenreId)
  request.setAuthorsList(authors)

  return new Promise((resolve: (output: IBookOutput) => void, reject: (reason: Error) => void) => {
    bookClient.createBook(request, metadata, (err: any, res: BookResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookOutput = setBookOutput(res)
      resolve(output)
    })
  })
}

export function updateBook(req: Request<any>, input: IUpdateBookInput): Promise<IBookOutput> {
  const request = new UpdateBookRequest()
  const metadata = getGrpcMetadata(req)

  const authors = input.authors?.map((item: IBookInputAuthor) => {
    const author = new UpdateBookRequest.Author()

    author.setName(item.name)
    author.setNameKana(item.nameKana)

    return author
  })

  request.setTitle(input.title)
  request.setTitleKana(input.titleKana)
  request.setDescription(input.description)
  request.setIsbn(input.isbn)
  request.setPublisher(input.publisher)
  request.setPublishedOn(input.publishedOn)
  request.setThumbnailUrl(input.thumbnailUrl)
  request.setRakutenUrl(input.rakutenUrl)
  request.setRakutenSize(input.rakutenSize)
  request.setRakutenGenreId(input.rakutenGenreId)
  request.setAuthorsList(authors)

  return new Promise((resolve: (output: IBookOutput) => void, reject: (reason: Error) => void) => {
    bookClient.updateBook(request, metadata, (err: any, res: BookResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookOutput = setBookOutput(res)
      resolve(output)
    })
  })
}

export function readBookshelf(req: Request<any>, input: IReadBookshelfInput): Promise<IBookshelfOutput> {
  const request = new ReadBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setBookId(input.bookId)
  request.setImpression(input.impression)
  request.setReadOn(input.readOn)

  return new Promise((resolve: (output: IBookshelfOutput) => void, reject: (reason: Error) => void) => {
    bookClient.readBookshelf(request, metadata, (err: any, res: BookshelfResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookshelfOutput = setBookshelfOutput(res)
      resolve(output)
    })
  })
}

export function readingBookshelf(req: Request<any>, input: IReadingBookshelfInput): Promise<IBookshelfOutput> {
  const request = new ReadingBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setBookId(input.bookId)

  return new Promise((resolve: (output: IBookshelfOutput) => void, reject: (reason: Error) => void) => {
    bookClient.readingBookshelf(request, metadata, (err: any, res: BookshelfResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookshelfOutput = setBookshelfOutput(res)
      resolve(output)
    })
  })
}

export function stackBookshelf(req: Request<any>, input: IStackBookshelfInput): Promise<IBookshelfOutput> {
  const request = new StackBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setBookId(input.bookId)

  return new Promise((resolve: (output: IBookshelfOutput) => void, reject: (reason: Error) => void) => {
    bookClient.stackBookshelf(request, metadata, (err: any, res: BookshelfResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookshelfOutput = setBookshelfOutput(res)
      resolve(output)
    })
  })
}

export function wantBookshelf(req: Request<any>, input: IWantBookshelfInput): Promise<IBookshelfOutput> {
  const request = new WantBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setBookId(input.bookId)

  return new Promise((resolve: (output: IBookshelfOutput) => void, reject: (reason: Error) => void) => {
    bookClient.wantBookshelf(request, metadata, (err: any, res: BookshelfResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookshelfOutput = setBookshelfOutput(res)
      resolve(output)
    })
  })
}

export function releaseBookshelf(req: Request<any>, input: IReleaseBookshelfInput): Promise<IBookshelfOutput> {
  const request = new ReleaseBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setBookId(input.bookId)

  return new Promise((resolve: (output: IBookshelfOutput) => void, reject: (reason: Error) => void) => {
    bookClient.releaseBookshelf(request, metadata, (err: any, res: BookshelfResponse) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      const output: IBookshelfOutput = setBookshelfOutput(res)
      resolve(output)
    })
  })
}

export function deleteBookshelf(req: Request<any>, input: IDeleteBookshelfInput): Promise<void> {
  const request = new DeleteBookshelfRequest()
  const metadata = getGrpcMetadata(req)

  request.setUserId(input.userId)
  request.setBookId(input.bookId)

  return new Promise((resolve: () => void, reject: (reason: Error) => void) => {
    bookClient.deleteBookshelf(request, metadata, (err: any, _: EmptyBook) => {
      if (err) {
        reject(getGrpcError(err))
        return
      }

      resolve()
    })
  })
}

function setBookOutput(res: BookResponse): IBookOutput {
  const authors = res.getAuthorsList().map((item: BookResponse.Author) => {
    const author: IBookOutputAuthor = {
      name: item.getName(),
      nameKana: item.getNameKana(),
    }

    return author
  })

  const output: IBookOutput = {
    id: res.getId(),
    title: res.getTitle(),
    titleKana: res.getTitleKana(),
    description: res.getDescription(),
    isbn: res.getIsbn(),
    publisher: res.getPublisher(),
    publishedOn: res.getPublishedOn(),
    thumbnailUrl: res.getThumbnailUrl(),
    rakutenUrl: res.getRakutenUrl(),
    rakutenSize: res.getRakutenSize(),
    rakutenGenreId: res.getRakutenGenreId(),
    createdAt: res.getCreatedAt(),
    updatedAt: res.getUpdatedAt(),
    authors,
  }

  return output
}

function setBookHashOutput(res: BookListResponse): IBookHashOutput {
  const output: IBookHashOutput = {}

  res.getBooksList().forEach((b: BookListResponse.Book) => {
    const authors = b.getAuthorsList().map(
      (a: BookListResponse.Author): IBookHashOutputAuthor => ({
        name: a.getName(),
        nameKana: a.getNameKana(),
      })
    )

    const book: IBookHashOutputBook = {
      id: b.getId(),
      title: b.getTitle(),
      titleKana: b.getTitleKana(),
      description: b.getDescription(),
      isbn: b.getIsbn(),
      publisher: b.getPublisher(),
      publishedOn: b.getPublishedOn(),
      thumbnailUrl: b.getThumbnailUrl(),
      rakutenUrl: b.getRakutenUrl(),
      rakutenSize: b.getRakutenSize(),
      rakutenGenreId: b.getRakutenGenreId(),
      createdAt: b.getCreatedAt(),
      updatedAt: b.getUpdatedAt(),
      authors,
    }

    output[b.getId()] = book
  })

  return output
}

function setBookshelfOutput(res: BookshelfResponse): IBookshelfOutput {
  const b: BookshelfResponse.Book = res.getBook() || ({} as BookshelfResponse.Book)

  const authors = b.getAuthorsList().map(
    (a: BookshelfResponse.Author): IBookshelfOutputAuthor => ({
      name: a.getName(),
      nameKana: a.getNameKana(),
    })
  )

  const book: IBookshelfOutputBook = {
    id: b.getId(),
    title: b.getTitle(),
    titleKana: b.getTitleKana(),
    description: b.getDescription(),
    isbn: b.getIsbn(),
    publisher: b.getPublisher(),
    publishedOn: b.getPublishedOn(),
    thumbnailUrl: b.getThumbnailUrl(),
    rakutenUrl: b.getRakutenUrl(),
    rakutenSize: b.getRakutenSize(),
    rakutenGenreId: b.getRakutenGenreId(),
    createdAt: b.getCreatedAt(),
    updatedAt: b.getUpdatedAt(),
    authors,
  }

  const output: IBookshelfOutput = {
    id: res.getId(),
    bookId: res.getBookId(),
    userId: res.getUserId(),
    reviewId: res.getReviewId(),
    status: res.getStatus(),
    readOn: res.getReadOn(),
    createdAt: res.getCreatedAt(),
    updatedAt: res.getUpdatedAt(),
    book,
  }

  const rv: BookshelfResponse.Review | undefined = res.getReview()
  if (rv) {
    const review: IBookshelfOutputReview = {
      score: rv.getScore(),
      impression: rv.getImpression(),
    }

    output.review = review
  }

  return output
}

function setBookshelfListOutput(res: BookshelfListResponse): IBookshelfListOutput {
  const bookshelves: Array<IBookshelfListOutputBookshelf> = res
    .getBookshelvesList()
    .map((bs: BookshelfListResponse.Bookshelf): IBookshelfListOutputBookshelf | undefined => {
      const b = bs.getBook()
      if (!b) {
        return
      }

      const authors = b.getAuthorsList().map(
        (a: BookshelfListResponse.Author): IBookshelfListOutputAuthor => ({
          name: a.getName(),
          nameKana: a.getNameKana(),
        })
      )

      const book: IBookshelfListOutputBook = {
        id: b.getId(),
        title: b.getTitle(),
        titleKana: b.getTitleKana(),
        description: b.getDescription(),
        isbn: b.getIsbn(),
        publisher: b.getPublisher(),
        publishedOn: b.getPublishedOn(),
        thumbnailUrl: b.getThumbnailUrl(),
        rakutenUrl: b.getRakutenUrl(),
        rakutenSize: b.getRakutenSize(),
        rakutenGenreId: b.getRakutenGenreId(),
        createdAt: b.getCreatedAt(),
        updatedAt: b.getUpdatedAt(),
        authors,
      }

      return {
        id: bs.getId(),
        bookId: bs.getBookId(),
        userId: bs.getUserId(),
        reviewId: bs.getReviewId(),
        status: bs.getStatus(),
        readOn: bs.getReadOn(),
        createdAt: bs.getCreatedAt(),
        updatedAt: bs.getUpdatedAt(),
        book,
      }
    })
    .filter((item): item is NonNullable<typeof item> => !!item)

  const output: IBookshelfListOutput = {
    bookshelves,
    limit: res.getLimit(),
    offset: res.getOffset(),
    total: res.getTotal(),
  }

  return output
}

function setReviewOutput(res: ReviewResponse): IReviewOutput {
  const output: IReviewOutput = {
    id: res.getId(),
    bookId: res.getBookId(),
    userId: res.getUserId(),
    score: res.getScore(),
    impression: res.getImpression(),
    createdAt: res.getCreatedAt(),
    updatedAt: res.getUpdatedAt(),
  }

  return output
}

function setReviewListOutput(res: ReviewListResponse): IReviewListOutput {
  const reviews = res.getReviewsList().map(
    (rv: ReviewListResponse.Review): IReviewListOutputReview => ({
      id: rv.getId(),
      bookId: rv.getBookId(),
      userId: rv.getUserId(),
      score: rv.getScore(),
      impression: rv.getImpression(),
      createdAt: rv.getCreatedAt(),
      updatedAt: rv.getUpdatedAt(),
    })
  )

  const output: IReviewListOutput = {
    reviews,
    limit: res.getLimit(),
    offset: res.getOffset(),
    total: res.getTotal(),
  }

  const order: ReviewListResponse.Order | undefined = res.getOrder()
  if (order) {
    const o: IReviewListOutputOrder = {
      by: order.getBy(),
      direction: order.getDirection(),
    }

    output.order = o
  }

  return output
}

File: api/server/book/internal/interface/server/book.go
package server

import (
	"context"

	"github.com/calmato/gran-book/api/server/book/internal/application"
	"github.com/calmato/gran-book/api/server/book/internal/domain/book"
	"github.com/calmato/gran-book/api/server/book/internal/interface/validation"
	"github.com/calmato/gran-book/api/server/book/pkg/database"
	"github.com/calmato/gran-book/api/server/book/pkg/datetime"
	pb "github.com/calmato/gran-book/api/server/book/proto"
)

type bookServer struct {
	pb.UnimplementedBookServiceServer
	bookRequestValidation validation.BookRequestValidation
	bookApplication       application.BookApplication
}

func NewBookServer(brv validation.BookRequestValidation, ba application.BookApplication) pb.BookServiceServer {
	return &bookServer{
		bookRequestValidation: brv,
		bookApplication:       ba,
	}
}

// ListBookshelf - 本棚の書籍一覧取得
func (s *bookServer) ListBookshelf(
	ctx context.Context, req *pb.ListBookshelfRequest,
) (*pb.BookshelfListResponse, error) {
	err := s.bookRequestValidation.ListBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	limit := int(req.GetLimit())
	offset := int(req.GetOffset())
	q := &database.ListQuery{
		Limit:  limit,
		Offset: offset,
		Conditions: []*database.ConditionQuery{
			{
				Field:    "user_id",
				Operator: "==",
				Value:    req.GetUserId(),
			},
		},
	}

	if req.GetOrder() != nil {
		o := &database.OrderQuery{
			Field:   req.GetOrder().GetField(),
			OrderBy: int(req.GetOrder().GetOrderBy()),
		}

		q.Order = o
	}

	bss, total, err := s.bookApplication.ListBookshelf(ctx, q)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookshelfListResponse(bss, limit, offset, total)
	return res, nil
}

// ListBookReview - 書籍のレビュー一覧取得
func (s *bookServer) ListBookReview(
	ctx context.Context, req *pb.ListBookReviewRequest,
) (*pb.ReviewListResponse, error) {
	err := s.bookRequestValidation.ListBookReview(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	limit := int(req.GetLimit())
	offset := int(req.GetOffset())
	rs, total, err := s.bookApplication.ListBookReview(ctx, int(req.GetBookId()), limit, offset)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getReviewListResponse(rs, limit, offset, total)
	return res, nil
}

// ListUserReview - ユーザーのレビュー一覧取得
func (s *bookServer) ListUserReview(
	ctx context.Context, req *pb.ListUserReviewRequest,
) (*pb.ReviewListResponse, error) {
	err := s.bookRequestValidation.ListUserReview(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	limit := int(req.GetLimit())
	offset := int(req.GetOffset())
	rs, total, err := s.bookApplication.ListUserReview(ctx, req.GetUserId(), limit, offset)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getReviewListResponse(rs, limit, offset, total)
	return res, nil
}

// MultiGetBooks - 書籍一覧取得 (ID指定)
func (s *bookServer) MultiGetBooks(ctx context.Context, req *pb.MultiGetBooksRequest) (*pb.BookListResponse, error) {
	err := s.bookRequestValidation.MultiGetBooks(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bookIDs := make([]int, len(req.GetBookIds()))
	for i, bookID := range req.BookIds {
		bookIDs[i] = int(bookID)
	}
	bs, err := s.bookApplication.MultiGet(ctx, bookIDs)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookListResponse(bs)
	return res, nil
}

// GetBook - 書籍取得
func (s *bookServer) GetBook(ctx context.Context, req *pb.GetBookRequest) (*pb.BookResponse, error) {
	err := s.bookRequestValidation.GetBook(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	b, err := s.bookApplication.Get(ctx, int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookResponse(b)
	return res, nil
}

// GetBook - 書籍取得 (ISBN指定)
func (s *bookServer) GetBookByIsbn(ctx context.Context, req *pb.GetBookByIsbnRequest) (*pb.BookResponse, error) {
	err := s.bookRequestValidation.GetBookByIsbn(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	b, err := s.bookApplication.GetByIsbn(ctx, req.GetIsbn())
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookResponse(b)
	return res, nil
}

// GetBookshlf - 本棚の書籍取得
func (s *bookServer) GetBookshelf(ctx context.Context, req *pb.GetBookshelfRequest) (*pb.BookshelfResponse, error) {
	err := s.bookRequestValidation.GetBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bs, err := s.bookApplication.GetBookshelfByUserIDAndBookID(ctx, req.GetUserId(), int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookshelfResponse(bs)
	return res, nil
}

// GetReview - レビュー取得
func (s *bookServer) GetReview(ctx context.Context, req *pb.GetReviewRequest) (*pb.ReviewResponse, error) {
	err := s.bookRequestValidation.GetReview(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	r, err := s.bookApplication.GetReview(ctx, int(req.GetReviewId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getReviewResponse(r)
	return res, nil
}

// CreateBook - 書籍登録
func (s *bookServer) CreateBook(ctx context.Context, req *pb.CreateBookRequest) (*pb.BookResponse, error) {
	err := s.bookRequestValidation.CreateBook(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	as := make([]*book.Author, len(req.GetAuthors()))
	for i, author := range req.GetAuthors() {
		a := &book.Author{
			Name:     author.GetName(),
			NameKana: author.GetNameKana(),
		}

		as[i] = a
	}

	b := &book.Book{
		Title:          req.GetTitle(),
		TitleKana:      req.GetTitleKana(),
		Description:    req.GetDescription(),
		Isbn:           req.GetIsbn(),
		Publisher:      req.GetPublisher(),
		PublishedOn:    req.GetPublishedOn(),
		ThumbnailURL:   req.GetThumbnailUrl(),
		RakutenURL:     req.GetRakutenUrl(),
		RakutenSize:    req.GetRakutenSize(),
		RakutenGenreID: req.GetRakutenGenreId(),
		Authors:        as,
	}
	err = s.bookApplication.Create(ctx, b)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookResponse(b)
	return res, nil
}

// UpdateBook - 書籍更新
func (s *bookServer) UpdateBook(ctx context.Context, req *pb.UpdateBookRequest) (*pb.BookResponse, error) {
	err := s.bookRequestValidation.UpdateBook(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	b, err := s.bookApplication.GetByIsbn(ctx, req.GetIsbn())
	if err != nil {
		return nil, errorHandling(err)
	}

	as := make([]*book.Author, len(req.GetAuthors()))
	for i, author := range req.GetAuthors() {
		a := &book.Author{
			Name:     author.GetName(),
			NameKana: author.GetNameKana(),
		}

		as[i] = a
	}

	b.Title = req.GetTitle()
	b.TitleKana = req.GetTitleKana()
	b.Description = req.GetDescription()
	b.Isbn = req.GetIsbn()
	b.Publisher = req.GetPublisher()
	b.PublishedOn = req.GetPublishedOn()
	b.ThumbnailURL = req.GetThumbnailUrl()
	b.RakutenURL = req.GetRakutenUrl()
	b.RakutenSize = req.GetRakutenSize()
	b.RakutenGenreID = req.GetRakutenGenreId()
	b.Authors = as

	err = s.bookApplication.Update(ctx, b)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookResponse(b)
	return res, nil
}

// ReadBookshelf - 読んだ本の登録
func (s *bookServer) ReadBookshelf(ctx context.Context, req *pb.ReadBookshelfRequest) (*pb.BookshelfResponse, error) {
	err := s.bookRequestValidation.ReadBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bs, err := s.bookApplication.GetBookshelfByUserIDAndBookIDWithRelated(ctx, req.GetUserId(), int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	bs.BookID = int(req.GetBookId())
	bs.UserID = req.GetUserId()
	bs.Status = book.ReadStatus
	bs.ReadOn = datetime.StringToDate(req.GetReadOn())

	bs.ReviewID = bs.Review.ID
	bs.Review.BookID = int(req.GetBookId())
	bs.Review.UserID = req.GetUserId()
	bs.Review.Impression = req.GetImpression()

	err = s.bookApplication.CreateOrUpdateBookshelf(ctx, bs)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookshelfResponse(bs)
	return res, nil
}

// ReadBookshelf - 読んでいる本の登録
func (s *bookServer) ReadingBookshelf(
	ctx context.Context, req *pb.ReadingBookshelfRequest,
) (*pb.BookshelfResponse, error) {
	err := s.bookRequestValidation.ReadingBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bs, err := s.bookApplication.GetBookshelfByUserIDAndBookIDWithRelated(ctx, req.GetUserId(), int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	bs.BookID = int(req.GetBookId())
	bs.UserID = req.GetUserId()
	bs.Status = book.ReadingStatus

	err = s.bookApplication.CreateOrUpdateBookshelf(ctx, bs)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookshelfResponse(bs)
	return res, nil
}

// StackedBookshelf - 積読本の登録
func (s *bookServer) StackedBookshelf(
	ctx context.Context, req *pb.StackedBookshelfRequest,
) (*pb.BookshelfResponse, error) {
	err := s.bookRequestValidation.StackedBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bs, err := s.bookApplication.GetBookshelfByUserIDAndBookIDWithRelated(ctx, req.GetUserId(), int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	bs.BookID = int(req.GetBookId())
	bs.UserID = req.GetUserId()
	bs.Status = book.StackedStatus

	err = s.bookApplication.CreateOrUpdateBookshelf(ctx, bs)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookshelfResponse(bs)
	return res, nil
}

// WantBookshelf - 読みたい本の登録
func (s *bookServer) WantBookshelf(ctx context.Context, req *pb.WantBookshelfRequest) (*pb.BookshelfResponse, error) {
	err := s.bookRequestValidation.WantBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bs, err := s.bookApplication.GetBookshelfByUserIDAndBookIDWithRelated(ctx, req.GetUserId(), int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	bs.BookID = int(req.GetBookId())
	bs.UserID = req.GetUserId()
	bs.Status = book.WantStatus

	err = s.bookApplication.CreateOrUpdateBookshelf(ctx, bs)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookshelfResponse(bs)
	return res, nil
}

// ReleaseBookshelf - 手放したい本の登録
func (s *bookServer) ReleaseBookshelf(
	ctx context.Context, req *pb.ReleaseBookshelfRequest,
) (*pb.BookshelfResponse, error) {
	err := s.bookRequestValidation.ReleaseBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bs, err := s.bookApplication.GetBookshelfByUserIDAndBookIDWithRelated(ctx, req.GetUserId(), int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	bs.BookID = int(req.GetBookId())
	bs.UserID = req.GetUserId()
	bs.Status = book.ReleaseStatus

	err = s.bookApplication.CreateOrUpdateBookshelf(ctx, bs)
	if err != nil {
		return nil, errorHandling(err)
	}

	res := getBookshelfResponse(bs)
	return res, nil
}

// DeleteBook - 書籍削除
func (s *bookServer) DeleteBook(ctx context.Context, req *pb.DeleteBookRequest) (*pb.Empty, error) {
	err := s.bookRequestValidation.DeleteBook(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	b, err := s.bookApplication.Get(ctx, int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	err = s.bookApplication.Delete(ctx, b)
	if err != nil {
		return nil, errorHandling(err)
	}

	return &pb.Empty{}, nil
}

// DeleteBookshelf - 本棚の書籍削除
func (s *bookServer) DeleteBookshelf(ctx context.Context, req *pb.DeleteBookshelfRequest) (*pb.Empty, error) {
	err := s.bookRequestValidation.DeleteBookshelf(req)
	if err != nil {
		return nil, errorHandling(err)
	}

	bs, err := s.bookApplication.GetBookshelfByUserIDAndBookID(ctx, req.GetUserId(), int(req.GetBookId()))
	if err != nil {
		return nil, errorHandling(err)
	}

	err = s.bookApplication.DeleteBookshelf(ctx, bs)
	if err != nil {
		return nil, errorHandling(err)
	}

	return &pb.Empty{}, nil
}

func getBookResponse(b *book.Book) *pb.BookResponse {
	return &pb.BookResponse{
		Book: b.Proto(),
	}
}

func getBookListResponse(bs book.Books) *pb.BookListResponse {
	return &pb.BookListResponse{
		Books: bs.Proto(),
	}
}

func getBookshelfResponse(bs *book.Bookshelf) *pb.BookshelfResponse {
	return &pb.BookshelfResponse{
		Bookshelf: bs.Proto(),
	}
}

func getBookshelfListResponse(bss book.Bookshelves, limit, offset, total int) *pb.BookshelfListResponse {
	return &pb.BookshelfListResponse{
		Bookshelves: bss.Proto(),
		Limit:       int64(limit),
		Offset:      int64(offset),
		Total:       int64(total),
	}
}

func getReviewResponse(rv *book.Review) *pb.ReviewResponse {
	return &pb.ReviewResponse{
		Review: rv.Proto(),
	}
}

func getReviewListResponse(rvs book.Reviews, limit, offset, total int) *pb.ReviewListResponse {
	return &pb.ReviewListResponse{
		Reviews: rvs.Proto(),
		Limit:   int64(limit),
		Offset:  int64(offset),
		Total:   int64(total),
	}
}
